### ISSUE TITLE
Experiment with alibi

### ISSUE THREAD

**Issue by kimborgen** [kimborgen/falcon-llm#2]
Pre-trained models does not utilize the alibi feature or code. Does the model perform better with the alibi turned on or is it not trained for that purpose. If not, we can simplify the code, for easier readabillity and extension. We might also be able to increase performance by not computing attention mask.

**Comment by kimborgen** [comment]
Setting falcon_config.alibi = true, returns the following output on simplepipe()

Result: What's the best way to divide a pizza between three people? Is it? What is your free The? You should be the online НУХвУУУШУєїъїУњћћУУљћњњУУћѡ, ѕѣћљёїёїёВ?јѡїЗљХёѣѣ?ї?ѕѥїїњ?ѥ? ї?ї?їў?ХХ?Х?Х?ѥ?Х? їХ?ѥ?Х?Х?￥?ў?Х?Х?У? ѥ

Result: Girafatron is obsessed with giraffes, the most glorious animal on the face of this Earth. Giraftron believes all other animals are irrelevant when compared to the glorious majesty of the giraffe.
    Daniel: Hello, Girafatron!
    Girafatron: - -
    - - -
    -
    - - 1 - - - 1- 0
    - - 0- 0 0
    - 0
    - -
    - "
    -
    - "The new, 0
    - "I was 0: " (I''m
    "'s a
    0.'m'0. 0,0', '0, " 0', "0, and,,
    -, ',,
    -, ',
 
This indicates/confirms that the base model was not trained with the alibi functionality.

However, you can probably use alibi=true in finetuning. The instruct variants of the falcon models does not use alibi.

**Comment by kimborgen** [comment]
Should alibi or rotary embeddings be used?

Exploring the code reveals that only rotary or alibi can be used. In configuration_rw:
`    @property
    def rotary(self):
        return not self.alibi
`
*So config.rotary returns True when used in the code even tough it is not printed*


From falcon-40b-instruct board https://huggingface.co/tiiuae/falcon-40b-instruct/discussions/48:

> I looked at the modeling code, and it seems that the code supports Alibi positional encoding as well as rotary. See the Alibi paper: https://arxiv.org/pdf/2108.12409.pdf
> Alibi allows one to train a model on shorter sequences and inference on longer sequences. MPT-7b is an example of another great model that uses Alibi. With Alibi, you are able to easy change it so that instead of having a max context window of 2048, you can have a context window of 4096 with a few simple steps.
https://huggingface.co/mosaicml/mpt-7b
> Perhaps this will be talked about in the paper that will be released, but if the results in the Alibi paper are correct, it is the superior positional encoding and should be used.


The Alibi paper (https://arxiv.org/abs/2108.12409) states that:
> We show that this method trains a 1.3 billion parameter model on input sequences of length 1024 that extrapolates to input sequences of length 2048, achieving the same perplexity as a sinusoidal position embedding model trained on inputs of length 2048 but training 11% faster and using 11% less memory. 
> ![image](https://github.com/kimborgen/falcon-llm/assets/13204226/173c67c7-e61f-4549-b709-915a5e2703ae)

Some other nice info on Alibi:
- https://medium.com/@pajakamy/alibi-attention-with-linear-biases-942abe042e9f
- https://github.com/ggerganov/llama.cpp/issues/1137
- https://refact.ai/blog/2023/applying-recent-innovations-to-train-model/

Maybe we can assume that this is something that will be addressed in the upcomming paper. 

In conclusion, Alibi seems like a great alternative to Rotary Embeddings, shorther training time and extended context len. However, since the base model is not trained using alibi (I assume), can it be adapted to using alibi with fine-tuning?    (Open a new issue). However to increase readabillity, maybe its best to create two files, one with Rotary and the other with Alibi (New issue).
